lib/mayaUsd/ufe/UsdRotateUndoableCommand.cpp

```diff
@@ -20,67 +20,61 @@
-	: Ufe::RotateUndoableCommand(item)
-	, fPrim(prim)
-	, fPath(ufePath)
-	, fNoRotateOp(false)
+TfToken UsdRotateUndoableCommand::rotXYZ("xformOp:rotateXYZ");
+
+UsdRotateUndoableCommand::UsdRotateUndoableCommand(
+    const UsdSceneItem::Ptr& item, double x, double y, double z)
+	: Ufe::RotateUndoableCommand(item),
+      UsdTRSUndoableCommandBase(item, x, y, z)
 {
-	// Since we want to change xformOp:rotateXYZ, and we need to store the prevRotate for 
-	// undo purpose, we need to make sure we convert it to common API xformOps (In case we have 
-	// rotateX, rotateY or rotateZ ops)
+	// Since we want to change xformOp:rotateXYZ, and we need to store the
+	// prevRotate for undo purposes, we need to make sure we convert it to
+	// common API xformOps (In case we have rotateX, rotateY or rotateZ ops)
 	try {
-		convertToCompatibleCommonAPI(prim);
+		convertToCompatibleCommonAPI(prim());
 	}
 	catch (...) {
-		// Since Maya cannot catch this error at this moment, store it until we actually rotate
+		// Since Maya cannot catch this error at this moment, store it until we
+		// actually rotate.
 		fFailedInit = std::current_exception(); // capture
-		return;
 	}
-
-	// Prim does not have a rotateXYZ attribute
-	const TfToken xrot("xformOp:rotateXYZ");
-	if (!fPrim.HasAttribute(xrot))
-	{
-		rotateOp(fPrim, fPath, 0, 0, 0);	// Add an empty rotate
-		fNoRotateOp = true;
-	}
-
-	fRotateAttrib = fPrim.GetAttribute(xrot);
-	fRotateAttrib.Get<GfVec3f>(&fPrevRotateValue);
 }
 
 UsdRotateUndoableCommand::~UsdRotateUndoableCommand()
-{
-}
+{}
 
 /*static*/
-UsdRotateUndoableCommand::Ptr UsdRotateUndoableCommand::create(const UsdPrim& prim, const Ufe::Path& ufePath, const Ufe::SceneItem::Ptr& item)
+UsdRotateUndoableCommand::Ptr UsdRotateUndoableCommand::create(
+    const UsdSceneItem::Ptr& item, double x, double y, double z)
 {
-	return std::make_shared<UsdRotateUndoableCommand>(prim, ufePath, item);
+	auto cmd = std::make_shared<MakeSharedEnabler<UsdRotateUndoableCommand>>(
+        item, x, y, z);
+    cmd->initialize();
+    return cmd;
 }
 
 void UsdRotateUndoableCommand::undo()
 {
 	// Check if initialization went ok.
 	if (!fFailedInit)
 	{
-		fRotateAttrib.Set(fPrevRotateValue);
+        UsdTRSUndoableCommandBase::undoImp();
 	}
-	// Todo : We would want to remove the xformOp
-	// (SD-06/07/2018) Haven't found a clean way to do it - would need to investigate
 }
 
 void UsdRotateUndoableCommand::redo()
 {
-	perform();
+    redoImp();
+}
+
+void UsdRotateUndoableCommand::addEmptyAttribute()
+{
+    performImp(0, 0, 0);	// Add an empty rotate
 }
 
-void UsdRotateUndoableCommand::perform()
+void UsdRotateUndoableCommand::performImp(double x, double y, double z)
 {
-	// No-op, use rotate to move the object
-	// The Maya rotate command directly invokes our rotate() method in its
-	// redoIt(), which is invoked both for the inital rotate and the redo.
+	rotateOp(prim(), path(), x, y, z);
```

Comment:
The fact that perform() was a no-op was a very incomplete implementation based on the behavior of the Maya move, rotate and scale commands, which perform undo themselves by simply setting the previous value.  For those uses, undo is not invoked.

When used in absolute mode, the parent command changes the local transform of an object.  Its undo should not be concerned about setting the previous values for the parented object's translate, rotate, and scale: the undo commands should do this themselves (separation of concerns).